tools/link.py

```diff
@@ -575,25 +575,92 @@ def include_and_export(name):
       settings.EXPORTED_RUNTIME_METHODS += ['ExitStatus']
 
 
+def set_initial_memory():
+  user_specified_initial_heap = 'INITIAL_HEAP' in user_settings
+
+  # INITIAL_HEAP cannot be used when the memory object is created in JS: we don't know
+  # the size of static data here and thus the total initial memory size.
+  if settings.IMPORTED_MEMORY:
+    if user_specified_initial_heap:
+      # Some of these could (and should) be implemented.
+      exit_with_error('INITIAL_HEAP is currently not compatible with IMPORTED_MEMORY (which is enabled indirectly via  SHARED_MEMORY, RELOCATABLE, ASYNCIFY_LAZY_LOAD_CODE)')
+    # The default for imported memory is to fall back to INITIAL_MEMORY.
+    settings.INITIAL_HEAP = -1
+
+  if not user_specified_initial_heap:
+    # For backwards compatibility, we will only use INITIAL_HEAP by default when the user
+    # specified neither INITIAL_MEMORY nor MAXIMUM_MEMORY. Both place an upper bounds on
+    # the overall initial linear memory (stack + static data + heap), and we do not know
+    # the size of static data at this stage. Setting any non-zero initial heap value in
+    # this scenario would risk pushing users over the limit they have set.
+    user_specified_initial = settings.INITIAL_MEMORY != -1
+    user_specified_maximum = 'MAXIMUM_MEMORY' in user_settings or 'WASM_MEM_MAX' in user_settings or 'BINARYEN_MEM_MAX' in user_settings
+    # INITIAL_HEAP does not allow us to cap maximum memory to initial memory without memory
+    # growth, which forces minimal WASM2JS builds to include "grow". Avoid regressing them.
```

Comment:
Can you explain this part.  What makes wasm2js different here?

Reply:
In wasm2js, there is an old optimization that skips including the "grow memory" function `if (initial_memory == maximum_memory)`. Of course, this is not a concern for native WASM.

Reply:
But shouldn't that optimization also kick in the growth is not allowed?

Perhaps we can address this issue so we don't need to have this rather obscure special case here.   Do you know where this "avoid grow memory" thing lives?  Is it in wasm2c itself?

Which test in particular regresses without this line?

Reply:
> But shouldn't that optimization also kick in the growth is not allowed?

I suspect I have got you confused with the wording of the comment... "INITIAL_HEAP does not allow us to cap maximum memory to initial memory without memory growth" <=> "INITIAL_HEAP does not allow us to cap maximum memory to initial memory, which try to do if memory growth is not allowed".

> Perhaps we can address this issue so we don't need to have this rather obscure special case here.

Maybe, but I would not know how at this point.

>  Do you know where this "avoid grow memory" thing lives?

Yep - it's this line: https://github.com/WebAssembly/binaryen/blob/feb8f24128c0c3bd53862b2f39acc5116f8ae87e/src/wasm2js.h#L458-L467, introduced in https://github.com/WebAssembly/binaryen/commit/7d1ff56acafae9c769bc8dd8da2c8ef3c66a2aa6.

> Which test in particular regresses without this line?

Any "hello world" WASM2JS, you can see the numbers in this commit - https://github.com/emscripten-core/emscripten/pull/21071/commits/47ec651b1349d5c8a54d0067dad20644ee4af761 (it's a pretty significant regression).

Reply:
If `ALLOW_MEMORY_GROWTH=0` then it should never be true that `wasm->memories[0]->max > wasm->memories[0]->initial` right?

Reply:
That's the issue - this is not true with `INITIAL_HEAP`. We can only cap if we know what to cap to.
```diff
  if not settings.ALLOW_MEMORY_GROWTH:
    if 'MAXIMUM_MEMORY' in user_settings:
      diagnostics.warning('unused-command-line-argument', 'MAXIMUM_MEMORY is only meaningful with ALLOW_MEMORY_GROWTH')
-   settings.MAXIMUM_MEMORY = settings.INITIAL_MEMORY
+   # Optimization: lower the default maximum memory to initial memory if possible.
+   if initial_memory_known:
+     settings.MAXIMUM_MEMORY = settings.INITIAL_MEMORY
```

Reply:
Are you saying that `ALLOW_MEMORY_GROWTH` is somehow not honored when `INITIAL_HEAP` is used?

If I just specify `INITIAL_HEAP` (i.e. the new default) what are the initial and max values for the memory in the resulting wasm binary?

Reply:
> Are you saying that ALLOW_MEMORY_GROWTH is somehow not honored when INITIAL_HEAP is used?

It is not honored insofar as the capping is not performed, but otherwise works just as before. I believe there is not much consequence to setting the WASM maximum memory field high as far as runtime performance is concerned.

> If I just specify INITIAL_HEAP (i.e. the new default) what are the initial and max values for the memory in the resulting wasm binary?

The initial memory will be "stack + static data + initial heap" (as you'd expect), the maximum memory will be 2GB (default).

Reply:
Resolved with the latest commits.